### PR TITLE
Fix hadoop batch ingest

### PR DESCRIPTION
**Pull request by kkrugler**
## Description
Numerous fixes and cleanup for the Hadoop batch segment generation code.
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
N/A

## Documentation
N/A

### CODE REVIEW

**Comment by kkrugler** [comment]
Fix for #6492

**Comment by kkrugler** [comment]
Hi @fx19880617 - I see that the Hadoop batch ingest unit test is failing, though it passes on my Mac when running `mvn clean test` from inside the `pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/` directory. What's the best way to view the build server test logs?

**Comment by fx19880617** [comment]
> Hi @fx19880617 - I see that the Hadoop batch ingest unit test is failing, though it passes on my Mac when running `mvn clean test` from inside the `pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/` directory. What's the best way to view the build server test logs?

will you be able to see the logs: https://github.com/apache/incubator-pinot/pull/6506/checks?check_run_id=1795216286

Seems that it's trying to copy a directory not a file?
```
[INFO] Running org.apache.pinot.plugin.ingestion.batch.hadoop.HadoopSegmentGenerationJobRunnerTest
Error:  Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.752 s <<< FAILURE! - in org.apache.pinot.plugin.ingestion.batch.hadoop.HadoopSegmentGenerationJobRunnerTest
Error:  org.apache.pinot.plugin.ingestion.batch.hadoop.HadoopSegmentGenerationJobRunnerTest.testSegmentGeneration  Time elapsed: 1.415 s  <<< FAILURE!
java.io.IOException: Source '/home/runner/work/incubator-pinot/incubator-pinot/pinot-plugins/pinot-input-format/pinot-csv/target/classes' exists but is a directory
	at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:1077)
	at org.apache.commons.io.FileUtils.copyFile(FileUtils.java:1038)
	at org.apache.pinot.plugin.ingestion.batch.hadoop.HadoopSegmentGenerationJobRunnerTest.testSegmentGeneration(HadoopSegmentGenerationJobRunnerTest.java:98)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:108)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:661)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:869)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1193)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:126)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:744)
	at org.testng.TestRunner.run(TestRunner.java:602)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:380)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:375)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1301)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1226)
	at org.testng.TestNG.runSuites(TestNG.java:1144)
	at org.testng.TestNG.run(TestNG.java:1115)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:136)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:112)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:99)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:145)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)

[INFO] 
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:    HadoopSegmentGenerationJobRunnerTest.testSegmentGeneration:98 » IO Source '/ho...
[INFO] 
Error:  Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
```

**Comment by kkrugler** [comment]
Hi @fx19880617 - it was a weird issue. Building locally, I would get a URI to a Pinot plugin jar file with:

``` java
SomePintoClassInPluginJar.class.getProtectionDomain().getCodeSource().getLocation().toURI()
```

But on the build server, this returns a URI to the directory containing the `SomePintoClassInPluginJar.class` file. I've changed the code to use a class from some dependency jar that's not part of Pinot (as we're just using it to test creating a tarball of files, and adding a dependency file to the distributed cache). So all checks seem to be passing now.

**Comment by fx19880617** [comment]
> Hi @fx19880617 - it was a weird issue. Building locally, I would get a URI to a Pinot plugin jar file with:
> 
> ```java
> SomePintoClassInPluginJar.class.getProtectionDomain().getCodeSource().getLocation().toURI()
> ```
> 
> But on the build server, this returns a URI to the directory containing the `SomePintoClassInPluginJar.class` file. I've changed the code to use a class from some dependency jar that's not part of Pinot (as we're just using it to test creating a tarball of files, and adding a dependency file to the distributed cache). So all checks seem to be passing now.

I think that's fine. We can just test packaging the jar and ensure the unpacked jar is the same.

**Review by fx19880617 — APPROVED** [review]
Overall lgtm! 
Many thanks for fixing this!!!

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java [hunk 1]
```diff
@@ -208,11 +220,12 @@ public void run()
     } else {
       LOGGER.info("Creating segments with data files: {}", filteredFiles);
       for (int i = 0; i < numDataFiles; i++) {
-        String dataFilePath = filteredFiles.get(i);
-
-        File localFile = new File("tmp");
+        // Typically PinotFS implementations list files without a protocol, so we lose (for example) the
+        // hdfs:// portion of the path. Call getFileURI() to fix this up.
+        URI inputFileURI = SegmentGenerationUtils.getFileURI(filteredFiles.get(i), inputDirURI);
```
[the remark this answers is not in the document]
Comment (kkrugler) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java:225]: It should be fixed in Standalone. I could file an issue for Spark, lmk.

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java [hunk 2]
```diff
@@ -269,33 +300,80 @@ public void run()
...
+   * @throws URISyntaxException 
+   */
+  private void moveFiles(PinotFS fs, URI sourceDir, URI destDir, boolean overwrite) throws IOException, URISyntaxException {
+    for (String sourcePath : fs.listFiles(sourceDir, true)) {
+      URI sourceFileUri = SegmentGenerationUtils.getFileURI(sourcePath, sourceDir);
+      String sourceFilename = FilenameUtils.getName(sourceFileUri.getPath());
+      URI destFileUri = SegmentGenerationUtils.getRelativeOutputPath(sourceDir, sourceFileUri, destDir).resolve(sourceFilename);
+      
+      if (!overwrite && fs.exists(destFileUri)) {
+        LOGGER.warn("Can't overwrite existing output segment tar file: {}", destFileUri);
```
[the remark this answers is not in the document]
Comment (kkrugler) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java:332]: Good idea, let me look into that

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (fx19880617) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java:225]: Yeah, I will make the changes for standalone and spark

[the remark this answers is not in the document]
Comment (kkrugler) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java:225]: Hi @fx19880617 I think I fixed this in Standalone a while back (that's what I meant by "It should be fixed in Standalone...", sorry for not being more clear.

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-common/src/test/java/org/apache/pinot/plugin/ingestion/batch/common/SegmentGenerationUtilsTest.java [hunk 3]
```diff
@@ -53,4 +53,31 @@ public void testRelativeURIs() throws URISyntaxException {
...
+
+    // S3 URI with userInfo (username/password)
+    validateFileURI(new URI("s3://username:password@bucket/path/to/"));
+  }
+
+  private void validateFileURI(URI directoryURI) throws URISyntaxException {
+    URI fileURI = new URI(directoryURI.toString() + "file");
+    String rawPath = fileURI.getRawPath();
+
+    Assert.assertEquals(SegmentGenerationUtils.getFileURI(rawPath, fileURI).toString(),
```
[the remark this answers is not in the document]
Comment (kkrugler) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-common/src/test/java/org/apache/pinot/plugin/ingestion/batch/common/SegmentGenerationUtilsTest.java:78]: Hi @fx19880617 - there was a test, for `file:/path/to/`. But thanks for asking, as I added more tests for different types of file URIs, and found an issue in the `getFileURI()` method with `file:///path/to/`. Plus I learned something about file URIs :)

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kkrugler) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java:332]: Unfortunately there's no way to do that before generating the segments, as the segment names (can) depend on the data used to generate the segment, and that's what we need to check for collision in the output directory.

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java [hunk 4]
```diff
@@ -269,33 +300,80 @@ public void run()
...
+        File pluginsTarGzFile = File.createTempFile("pinot-plugins-", ".tar.gz");
         TarGzCompressionUtils.createTarGzFile(pluginsRootDir, pluginsTarGzFile);
-      } catch (IOException e) {
+        
+        // Copy to staging directory
+        Path cachedPluginsTarball = new Path(stagingDirURI.toString(), SegmentGenerationUtils.PINOT_PLUGINS_TAR_GZ);
+        outputDirFS.copyFromLocalFile(pluginsTarGzFile, cachedPluginsTarball.toUri());
+        job.addCacheFile(cachedPluginsTarball.toUri());
+      } catch (Exception e) {
         LOGGER.error("Failed to tar plugins directory", e);
```
[the remark this answers is not in the document]
Comment (kkrugler) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java]: Done

**Review by fx19880617 — APPROVED** [review]

**Comment by fx19880617** [comment]
Thanks for fixing this!

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (fx19880617) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java:224]: Cool! Making a PR for Spark as well: https://github.com/apache/incubator-pinot/pull/6537

**Pull request merged**